### ISSUE TITLE
Issue or user error?

### ISSUE THREAD

**Issue by ericphanson** [jkrumbiegel/ReadableRegex.jl#3]
```julia
julia> one(x) = exactly(1, x)
one (generic function with 1 method)

julia> find_graphics_with_arg = @compile one("\\includegraphics[") * capture(zero_or_more(NON_LINEBREAK)) * one("]{") * capture(zero_or_more(NON_LINEBREAK)) * one("}")
r"(.*)(.*)"
```
I did not expect this to result in two anythings in capture groups! Wanted to get the args and path from a latex `\includegraphics` command. Does `exactly` not do what I think it should here? Or is it a bug?

**Comment by ericphanson** [comment]
Ah, it's some kind of macro escaping issue:
```julia
julia> _one(x) = exactly(1, x)
_one (generic function with 1 method)

julia> find_graphics_with_arg = @compile _one("\\includegraphics[") * capture(zero_or_more(NON_LINEBREAK)) * _one("]{") * capture(zero_or_more(NON_LINEBREAK)) * _one("}")
ERROR: UndefVarError: _one not defined
```
(so the `one` from the post is `Base.one` instead of `Main.one`, which is why we get the results I see there).

**Comment by jkrumbiegel** [comment]
Yes the problem with compile is that it cannot call your functions, because it directly constructs the regex object and places that in the code, like the `r""` string macro (I just looked at that implementation and copied that, I think). Compare this to a macro that splices function definitions into the source code, these can later be called. At least that's how I understood this limitation.

So all the compiled stuff needs to be in one expression, but I guess you could do functions if you defined them within `@compile begin ... end`?

Although, your actual issue is different :) Because you are trying to define a convenience function which you don't need. `exactly(1, x)` is equivalent to just concatenating a string into your `RegexString` (x is automatically escaped if it contains special regex characters). So you can just write:

```julia
find_graphics_with_arg = @compile "\\includegraphics[" * capture(zero_or_more(NON_LINEBREAK)) * "]{" * capture(zero_or_more(NON_LINEBREAK)) * "}"
```

**Comment by ericphanson** [comment]
> Although, your actual issue is different :) Because you are trying to define a convenience function which you don't need. exactly(1, x) is equivalent to just concatenating a string into your RegexString (x is automatically escaped if it contains special regex characters). 

Oh! That's great, thanks for pointing that out.

> Yes the problem with compile is that it cannot call your functions, because it directly constructs the regex object and places that in the code, like the r"" string macro (I just looked at that implementation and copied that, I think). Compare this to a macro that splices function definitions into the source code, these can later be called. At least that's how I understood this limitation.

So the macro in base just does
```julia
macro r_str(pattern, flags...) Regex(pattern, flags...) end
```
while in ReadableRegex it's
```julia
macro compile(exp)
    regexstring = @eval $exp
    Regex(regexstring.s)
end
```
I think it's the `@eval` that's the problem, since that does not happen in the local scope, and that doesn't seem to be there in the Base macro.

I don't really understand enough about macros to know why the Base one works or what should be done differently, if anything.
